### PR TITLE
Turtles - Shari

### PR DESCRIPTION
**Pull request by sharifedak12**
Passes all tests - plus a bonus round!

### CODE REVIEW

File: swap_meet/item.py [hunk 1]
```diff
@@ -1,2 +1,22 @@
...
+        if self.condition in range(1, 2):
+            return "signs of wear"
+        if self.condition in range(2, 3):
+            return "lightly used"
+        if self.condition in range(3, 4):
+            return "nearly new"
+        if self.condition in range(4, 5):
+            return "almost flawless"
+        if self.condition == 5:
+            return "mint!"
```
Comment (kelsey-steven-ada) [inline swap_meet/item.py:22]: I don't think the `range` calls here are doing what you intend. `range` is going to return a sequence of whole numbers from the start inclusive to the end exclusive, meaning `range(1, 2)` will return a sequence with only `1` in it. 

If we want to account for float values, the if-statements could be changed to compare if `condition` is less than a particular value:
```python
        if self.condition < 1:
            return "heavily used"
        if self.condition < 2:
            return "signs of wear"
        if self.condition < 3:
            return "lightly used"
        if self.condition < 4:
            return "nearly new"
        if self.condition < 5:
            return "almost flawless"
        if self.condition == 5:
            return "mint!"
```

Another approach for finding the condition description could be to make a dictionary that maps conditions as keys to descriptions as values. There's a trade off of using more memory creating the map variable, but if memory isn't a concern, for me, it's a little easier to read.
```python
        condition_map = {
            0: "heavily used.",
            1: "signs of wear",
            2: "lightly used",
            3: "nearly new",
            4: "almost flawless",
            5: "mint!"
        }
        
        condition_rounded_down  = int(self.condition)
        return condition_map[condition_rounded_down]
```

File: swap_meet/vendor.py [hunk 2]
```diff
@@ -1,2 +1,73 @@
...
+    def get_newest_item(self):
+        newest_age = min(item.age for item in self.inventory)
+        for item in self.inventory:
+            if item.age is newest_age:
+                return item
+
+    def swap_newest(self, other):
+        self_newest_item = self.get_newest_item()
+        their_newest_item = other.get_newest_item()
+        self.swap_items(other, self_newest_item, their_newest_item)
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:73]: Nice improvements & use of helper functions for them!

File: swap_meet/vendor.py [hunk 3]
```diff
@@ -1,2 +1,73 @@
...
+            return False
+        else:
+            self.swap_items(other, self_best, their_best)
+            return True
+
+    def get_newest_item(self):
+        newest_age = min(item.age for item in self.inventory)
+        for item in self.inventory:
+            if item.age is newest_age:
+                return item
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:68]: This approach is solid, but we make 2 passes over the data: one using `min` to get the newest age, and one to match it up with an object. There's many ways we could approach this to make a single pass over the data. 
One options is to forgo the `min` function and find the min ourselves in a single loop:
```python
if not self.inventory:
    return None

newest_item = self.inventory[0]
for item in self.inventory:
    if item.age < newest_item.age:
         newest_item = item
return newest_item
``` 
But we could use `min` along with a lambda function that gets the `age` for each item:
```python
return min(self.inventory, key=lambda item: item.age)
```

File: swap_meet/electronics.py [hunk 4]
```diff
@@ -1,2 +1,9 @@
-class Electronics:
-    pass
+from swap_meet.item import Item
+
+
+class Electronics(Item):
```
Comment (kelsey-steven-ada) [inline swap_meet/electronics.py:4]: The subclasses look great!

File: swap_meet/vendor.py [hunk 5]
```diff
@@ -1,2 +1,73 @@
 class Vendor:
-    pass
\ No newline at end of file
+    def __init__(self, inventory=None):
+        self.inventory = inventory if inventory is not None else []
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:3]: Great use of a Ternary operator to simplify handling the `None` default value!

File: swap_meet/vendor.py [hunk 6]
```diff
@@ -1,2 +1,73 @@
...
+    def add(self, item):
+        self.inventory.append(item)
+        return item
+
+    def remove(self, item):
+        if item in self.inventory:
+            self.inventory.remove(item)
+            return item
+        else:
+            return False
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:14]: I like inverting the logic of the condition to be able to treat it like a guard clause, which allows us to outdent the main logic of the function something like the following:
```python
        if item not in self.inventory:
            return False

        self.inventory.remove(item)
        return item
```

Another approach we could take is to try to remove the item directly, and handle the `ValueError` that occurs if it's not there, and return `False` to handle it.

File: swap_meet/vendor.py [hunk 7]
```diff
@@ -1,2 +1,73 @@
...
+
+    def get_by_category(self, category):
+        filtered_items = []
+        for item in self.inventory:
+            if item.category == category:
+                filtered_items.append(item)
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:20]: Notice this pattern of: 
```python
result_list = []
for element in source_list:
    if some_condition(element):
        result_list.append(element)
```

can be rewritten using a list comprehension as:
```python
result_list = [element for element in source_list if some_condition(element)]
```

Which here would look like:
```python
filtered_items = [item for item in self.inventory if item.category == category]
```

At first, this may seem more difficult to read, but comprehensions are a very common python style, so try getting used to working with them!

File: swap_meet/vendor.py [hunk 8]
```diff
@@ -1,2 +1,73 @@
...
+            if their_item in vendor.inventory:
+                self.remove(my_item)
+                self.add(their_item)
+                vendor.remove(their_item)
+                vendor.add(my_item)
+                return True
+            else:
+                return False
+        else:
+            return False
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:34]: We could simplify the if-statements a bit if we do our valid data check at the top of the function as a guard statement:
```suggestion
        if my_item not in self.inventory or their_item not in friend.inventory:
            return False
            
        self.remove(my_item)
        self.add(their_item)
        vendor.remove(their_item)
        vendor.add(my_item)
        return True
```

File: swap_meet/vendor.py [hunk 9]
```diff
@@ -1,2 +1,73 @@
...
+
+    def get_best_by_category(self, category):
+        items = self.get_by_category(category)
+        if not items:
+            return None
+        else:
+            best_rating = max(item.condition for item in items)
+            for item in items:
+                if item.condition is best_rating:
+                    return item
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:53]: Great choice to use the `max` function! We could remove the loop here if we use the max function on the list of items and pass a lambda function that tells it to compare based on each item's condition:
```suggestion
            return max(items, lambda item: item.condition)
```
Another approach would be to sort the list of items by condition so we know where the item with the best condition is:
```suggestion
            sorted_list = sorted(items, key=lambda item: item.condition)
            return sorted_list[-1]
```

File: swap_meet/vendor.py [hunk 10]
```diff
@@ -1,2 +1,73 @@
...
+            return False
+
+    def get_by_category(self, category):
+        filtered_items = []
+        for item in self.inventory:
+            if item.category == category:
+                filtered_items.append(item)
+        return filtered_items
+
+    def swap_items(self, vendor, my_item, their_item):
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:23]: I would consider renaming parameters called `vendor` across the file. To avoid potential confusion, I generally try to avoid using the name of a class as a variable name within that class. Here we could call `vendor` something like `friend_vendor` or `other_vendor` that is a little more distinct.

File: swap_meet/vendor.py [hunk 11]
```diff
@@ -1,2 +1,73 @@
...
+            else:
+                return False
+        else:
+            return False
+
+    def swap_first_item(self, vendor):
+        try:
+            self_first_item = self.inventory[0]
+            vendor_first_item = vendor.inventory[0]
+            self.swap_items(vendor, self_first_item, vendor_first_item)
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:40]: Nice code reuse!

Something to consider: by using `swap_items`, this winds up being O(n) (mostly due to removing the items from the inventory). If it was important to lower the Big O complexity, we could swap the items in place here to have O(1) runtime with something like:
```python
self.inventory[0], vendor.inventory[0] = vendor.inventory[0], self.inventory[0]
```

File: swap_meet/vendor.py [hunk 12]
```diff
@@ -1,2 +1,73 @@
...
+                vendor.remove(their_item)
+                vendor.add(my_item)
+                return True
+            else:
+                return False
+        else:
+            return False
+
+    def swap_first_item(self, vendor):
+        try:
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:37]: Nice approach to check for invalid data and get the items to swap in a single step!

File: swap_meet/vendor.py (hunk 9 above)
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:53]: Similar to the feedback in the remove function, we could remove the else:
```suggestion

        best_rating = max(item.condition for item in items)
        for item in items:
            if item.condition is best_rating:
                return item
```

File: swap_meet/vendor.py [hunk 13]
```diff
@@ -1,2 +1,73 @@
...
+                    return item
+
+    def swap_best_by_category(self, other, my_priority, their_priority):
+        self_best = self.get_best_by_category(their_priority)
+        their_best = other.get_best_by_category(my_priority)
+        if their_best is None or self_best is None:
+            return False
+        else:
+            self.swap_items(other, self_best, their_best)
+            return True
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:62]:
```suggestion

        self.swap_items(other, self_best, their_best)
        return True
```

File: swap_meet/vendor.py [hunk 14]
```diff
@@ -1,2 +1,73 @@
...
+        items = self.get_by_category(category)
+        if not items:
+            return None
+        else:
+            best_rating = max(item.condition for item in items)
+            for item in items:
+                if item.condition is best_rating:
+                    return item
+
+    def swap_best_by_category(self, other, my_priority, their_priority):
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:55]: Great approach to this function!

File: tests/unit_tests/test_wave_01.py [hunk 15]
```diff
@@ -49,7 +49,8 @@ def test_removing_not_found_is_false():
 
     result = vendor.remove(item)
 
-    raise Exception("Complete this test according to comments below.")
+    #raise Exception("Complete this test according to comments below.")
     # *********************************************************************
     # ****** Complete Assert Portion of this test **********
     # *********************************************************************
+    assert result == False
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_01.py:56]: This assert is great for confirming the return value is correct, but what else could we assert to make sure there were no side affects like changes to the inventory?

File: tests/unit_tests/test_wave_02.py [hunk 16]
```diff
@@ -34,7 +34,8 @@ def test_get_no_matching_items_by_category():
 
     items = vendor.get_by_category("electronics")
 
-    raise Exception("Complete this test according to comments below.")
+    
     # *********************************************************************
     # ****** Complete Assert Portion of this test **********
     # *********************************************************************
+    assert items == []
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_02.py:41]: The feedback in test_wave_01 applies here as well.

File: tests/unit_tests/test_wave_06.py [hunk 17]
```diff
@@ -76,16 +76,18 @@ def test_swap_best_by_category():
...
     # ****** Complete Assert Portion of this test **********
     # *********************************************************************
     # Assertions should check:
     # - That the results is truthy
     # - That tai and jesse's inventories are the correct length
     # - That all the correct items are in tai and jesse's inventories, including the items which were swapped from one vendor to the other
+    assert len(tai.inventory) == 3
+    assert len(jesse.inventory) == 3
+    assert tai.inventory == [item_a, item_b, item_f]
+    assert jesse.inventory == [item_d, item_e, item_c]
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_06.py:89]: Nice assertions for all the relevant data! How could we check the right values are in the right list if we could not guarantee the order of the lists? This feedback applies to all the tests below where a vendor's inventory is compared to a list with a specific ordering.

File: tests/unit_tests/test_wave_07 - BONUS ROUND.py [hunk 18]
```diff
@@ -0,0 +1,50 @@
+import pytest
+from swap_meet.vendor import Vendor
+from swap_meet.clothing import Clothing
+from swap_meet.decor import Decor
+from swap_meet.electronics import Electronics
+
+
+def test_newest():
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_07 - BONUS ROUND.py:8]: Solid happy-path tests! What other tests could we add to cover edge cases around the age attribute and age operations?

**Review by kelsey-steven-ada — COMMENTED** [review]
🎉 You completed Unit 1! 🎉 Great work on this project, I've left some suggestions for refactoring. Take a look, and let me know if anything needs clarification!
